### ISSUE TITLE
False-negative allows to return any Collection.

### ISSUE THREAD

**Issue by biozshock** [phpstan/phpstan-doctrine#434]
The code below does not show any errors. 
However if you copy all classes to the playground the error is detected. https://phpstan.org/r/4529cd8b-36d6-459d-b737-8fe3c48936f0

```php

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;

class Foo{}

class Bar {
	
	/** @var Collection<int, Foo> */
	private Collection $collection;
	
	public function __construct()
	{
		$this->collection = new ArrayCollection();
	}
	
	/**
	 * @return Collection<int, Foo>
	 */
	function get(): Collection
	{
		$col = new ArrayCollection();
		$col->add(new \stdClass);

		return $col;
	}
}
```

Thanks for making PHPStan!

**Comment by ondrejmirtes** [comment]
So what’s your expectation? Do you expect your application to also report this error but it currently doesn’t when analyzed? This is kind of missing from your description.

Can you share your config and how you run PHPStan on the command line?

**Comment by biozshock** [comment]
Oh, sorry. Missed that part.

Yes, i think PHPStan should report something like `Method Bar::get() should return Collection<int, Foo> but returns ArrayCollection<int, stdClass>.`.

The config is simple:
```
includes:
	- vendor/phpstan/phpstan-deprecation-rules/rules.neon
	- vendor/phpstan/phpstan-strict-rules/rules.neon
	- vendor/phpstan/phpstan-symfony/extension.neon
	- vendor/phpstan/phpstan-symfony/rules.neon
	- vendor/phpstan/phpstan-doctrine/extension.neon
	- vendor/phpstan/phpstan-doctrine/rules.neon

parameters:
    level: max
    treatPhpDocTypesAsCertain: false
    tmpDir: var/cache/dev
    symfony:
        container_xml_path: var/cache/dev/App_KernelDevDebugContainer.xml
        console_application_loader: config/phpstan/console-loader.php
    doctrine:
        repositoryClass: App\Repository\AbstractRepository
        objectManagerLoader: config/phpstan/doctrine-orm-bootstrap.php
``` 

console-loader.php:
```php
<?php declare(strict_types = 1);

use App\Kernel;
use Symfony\Bundle\FrameworkBundle\Console\Application;

require dirname(__DIR__) . '/bootstrap.php';
$kernel = new Kernel($_SERVER['APP_ENV'], (bool) $_SERVER['APP_DEBUG']);
return new Application($kernel);
```

doctrine-orm-bootstrap.php
```php
<?php declare(strict_types = 1);

use App\Kernel;

require dirname(__DIR__) . '/bootstrap.php';
$kernel = new Kernel($_SERVER['APP_ENV'], (bool) $_SERVER['APP_DEBUG']);
$kernel->boot();
return $kernel->getContainer()->get('doctrine')->getManager();
```

Command: `phpstan analyse -c phpstan.neon src -l max --memory-limit=1G`, where `phpstan` is a link to `vendor/bin/phpstan`

**Comment by ondrejmirtes** [comment]
Everything looks alright. Please create a small repository that shows the problem you’re having.

**Comment by biozshock** [comment]
Done. https://github.com/biozshock/phpstan-434 . GH Action run: https://github.com/biozshock/phpstan-434/actions/runs/4402761684/jobs/7710291794

Edit: line with an error that's not detected: https://github.com/biozshock/phpstan-434/blob/main/src/Entity/Bar.php#L26

**Comment by ondrejmirtes** [comment]
This is a truthful reproduction of your problem: https://phpstan.org/r/bf1da1d9-5d99-4845-a0ae-6e368367e466 (the constructor was missing).

Your code in the project also reports `ArrayCollection<*NEVER*, *NEVER*>`. Which means that the collection doesn't know it's no longer empty. If we let PHPStan know about this using `@phpstan-self-out` annotation, it suddenly knows (alongside with some errors we don't want): https://phpstan.org/r/58dfadf5-8f88-4749-be6b-215c96237d80

You might want to suggest this annotation in doctrine/collections (but I'm not sure how received it's going to be there).

Related: https://github.com/phpstan/phpstan/issues/6732
